src/Microsoft.AspNetCore.BuildTools.Tasks/GetGitCommitInfo.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using System.Linq;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.AspNetCore.BuildTools
+{
+    public class GetGitCommitInfo : Task
+    {
+        /// <summary>
+        /// A folder inside the git project. Does not need to be the top folder.
+        /// This task will search upwards for the .git folder.
+        /// </summary>
+        [Required]
+        public string WorkingDirectory { get; set; }
+
+        /// <summary>
+        /// The name of the branch HEAD is pointed to. Can be null or empty
+        /// for repositories in detached HEAD mode.
+        /// </summary>
+        [Output]
+        public string Branch { get; set; }
+
+        /// <summary>
+        /// The full commit SHA of the current commit referenced by HEAD.
+        /// </summary>
+        [Output]
+        public string CommitHash { get; set; }
+
+        /// <summary>
+        /// The folder containing the '.git', not the .git folder itself.
+        /// </summary>
+        [Output]
+        public string GitProjectRootPath { get; set; }
+
+        public override bool Execute()
+        {
+            Branch = null;
+            CommitHash = null;
+
+            var gitRoot = new DirectoryInfo(WorkingDirectory);
+            while (gitRoot != null)
+            {
+                if (gitRoot.GetDirectories(".git").Any())
+                {
+                    break;
+                }
+                gitRoot = gitRoot.Parent;
+            }
+
+            if (gitRoot == null)
+            {
+                Log.LogError("Could not find the git directory for '{0}'", WorkingDirectory);
+                return false;
+            }
+
+            GitProjectRootPath = gitRoot.FullName;
+
+            var headFile = Path.Combine(GitProjectRootPath, ".git", "HEAD");
+            if (!File.Exists(headFile))
+            {
+                Log.LogError("Unable to determine active git branch.");
+                return false;
+            }
+
+            var content = File.ReadAllText(headFile).Trim();
+            if (!content.StartsWith("ref: refs/heads/"))
+            {
+                Log.LogError("'.git/HEAD' file in unexpected format, unable to determine active git branch");
```

Comment:
Let's include the `content` as part of the error. Would be easy to diagnose if this ever fails.